File: frontend/src/utils/ifrsDotPathTranslate.ts
/**
 * Translate CompanyOnboarding dot-path keys → ifrs_line_item_master triples.
 * Keep backend/app/data/ifrs_dot_path_map.json in sync with src/data/ifrs_dot_path_map.json
 */
import mapData from "../data/ifrs_dot_path_map.json";

export type MatchQuality = "exact" | "best_fit" | "unmapped" | "unknown";

export type MasterTriple = {
  ifrs_statement: string;
  ifrs_section: string;
  ifrs_line_item: string;
  match_quality: MatchQuality;
  note?: string;
};

export type TranslateResult =
  | { ok: true; dot_path: string; triple: MasterTriple }
  | { ok: false; dot_path: string; match_quality: MatchQuality; note?: string; error: string };

const MAPPINGS = mapData.mappings as Record<
  string,
  {
    ifrs_statement: string | null;
    ifrs_section: string | null;
    ifrs_line_item: string | null;
    match_quality: MatchQuality;
    note?: string;
  }
>;

export function translateDotPath(dotPath: string): TranslateResult {
  const raw = (dotPath || "").trim();
  const entry = MAPPINGS[raw];
  if (!entry) {
    console.warn(`ifrsDotPathTranslate: unknown dot-path ${raw}`);
    return { ok: false, dot_path: raw, match_quality: "unknown", error: `Unknown dot-path: ${raw}` };
  }
  if (entry.match_quality === "unmapped" || !entry.ifrs_line_item) {
    console.warn(`ifrsDotPathTranslate: unmapped dot-path ${raw}`, entry.note);
    return {
      ok: false,
      dot_path: raw,
      match_quality: "unmapped",
      note: entry.note,
      error: entry.note || "No master line item mapping",
    };
  }
  if (entry.match_quality === "best_fit") {
    console.info(`ifrsDotPathTranslate: best_fit ${raw} → ${entry.ifrs_line_item}`, entry.note);
  }
  return {
    ok: true,
    dot_path: raw,
    triple: {
      ifrs_statement: entry.ifrs_statement!,
      ifrs_section: entry.ifrs_section!,
      ifrs_line_item: entry.ifrs_line_item!,
      match_quality: entry.match_quality,
      note: entry.note,
    },
  };
}

export type CoaTemplateEntry = {
  gl_code: string;
  gl_description: string;
  ifrs_statement: string;
  ifrs_section: string;
  ifrs_line_item: string;
  dot_path?: string;
  match_quality?: MatchQuality;
};

export function translateOnboardingMappings(
  chartOfAccounts: Array<{ glCode: string; accountName: string }>,
  mappings: Record<string, string>
): {
  entries: CoaTemplateEntry[];
  failures: TranslateResult[];
  skippedGlCodes: Array<{ glCode: string; accountName: string; reason: string }>;
} {
  const entries: CoaTemplateEntry[] = [];
  const failures: TranslateResult[] = [];
  const skippedGlCodes: Array<{ glCode: string; accountName: string; reason: string }> = [];
  for (const row of chartOfAccounts) {
    const dotPath = mappings[row.glCode];
    if (!dotPath) {
      skippedGlCodes.push({
        glCode: row.glCode,
        accountName: row.accountName,
        reason: "No IFRS mapping selected for this GL code",
      });
      continue;
    }
    const tr = translateDotPath(dotPath);
    if (!tr.ok) {
      failures.push(tr);
      continue;
    }
    entries.push({
      gl_code: row.glCode,
      gl_description: row.accountName,
      ifrs_statement: tr.triple.ifrs_statement,
      ifrs_section: tr.triple.ifrs_section,
      ifrs_line_item: tr.triple.ifrs_line_item,
      dot_path: dotPath,
      match_quality: tr.triple.match_quality,
    });
  }
  return { entries, failures, skippedGlCodes };
}

export function listUnmappedDotPaths(): string[] {
  return Object.entries(MAPPINGS)
    .filter(([, v]) => v.match_quality === "unmapped")
    .map(([k]) => k);
}

File: backend/app/services/ifrs_dot_path_translate.py
"""
Translate CompanyOnboarding dot-path keys → ifrs_line_item_master triples.

Source of truth: backend/app/data/ifrs_dot_path_map.json
(keep in sync with frontend/src/data/ifrs_dot_path_map.json)
"""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, TypedDict

logger = logging.getLogger(__name__)

_MAP_PATH = Path(__file__).resolve().parent.parent / "data" / "ifrs_dot_path_map.json"


class MasterTriple(TypedDict, total=False):
    ifrs_statement: str
    ifrs_section: str
    ifrs_line_item: str
    match_quality: str
    note: str


class TranslateResult(TypedDict):
    ok: bool
    dot_path: str
    ifrs_statement: str | None
    ifrs_section: str | None
    ifrs_line_item: str | None
    match_quality: str
    note: str | None
    error: str | None


def _load_map() -> dict[str, dict[str, Any]]:
    with _MAP_PATH.open(encoding="utf-8") as f:
        data = json.load(f)
    return dict(data.get("mappings") or {})


def translate_dot_path(dot_path: str) -> TranslateResult:
    """Map one dot-path to master triple. Unmapped paths return ok=False."""
    raw = (dot_path or "").strip()
    entry = _load_map().get(raw)
    if not entry:
        logger.warning("ifrs_dot_path_translate: unknown dot-path %r", raw)
        return {
            "ok": False,
            "dot_path": raw,
            "ifrs_statement": None,
            "ifrs_section": None,
            "ifrs_line_item": None,
            "match_quality": "unknown",
            "note": None,
            "error": f"Unknown dot-path: {raw}",
        }
    quality = str(entry.get("match_quality") or "")
    if quality == "unmapped" or not entry.get("ifrs_line_item"):
        logger.warning(
            "ifrs_dot_path_translate: unmapped dot-path %r — %s",
            raw,
            entry.get("note"),
        )
        return {
            "ok": False,
            "dot_path": raw,
            "ifrs_statement": None,
            "ifrs_section": None,
            "ifrs_line_item": None,
            "match_quality": "unmapped",
            "note": entry.get("note"),
            "error": entry.get("note") or "No master line item mapping",
        }
    if quality == "best_fit":
        logger.info(
            "ifrs_dot_path_translate: best_fit %r → %s / %s — %s",
            raw,
            entry.get("ifrs_section"),
            entry.get("ifrs_line_item"),
            entry.get("note"),
        )
    return {
        "ok": True,
        "dot_path": raw,
        "ifrs_statement": entry["ifrs_statement"],
        "ifrs_section": entry["ifrs_section"],
        "ifrs_line_item": entry["ifrs_line_item"],
        "match_quality": quality,
        "note": entry.get("note"),
        "error": None,
    }


def translate_dot_path_mappings(
    mappings: dict[str, str],
) -> tuple[list[dict[str, Any]], list[TranslateResult]]:
    """
    Convert { gl_code: dot_path } onboarding map to template entries.
    Returns (successful_entries, failures_per_gl).
    """
    successes: list[dict[str, Any]] = []
    failures: list[TranslateResult] = []
    for gl_code, dot_path in mappings.items():
        tr = translate_dot_path(dot_path)
        if not tr["ok"]:
            failures.append(tr)
            continue
        successes.append(
            {
                "gl_code": gl_code,
                "gl_description": gl_code,
                "ifrs_statement": tr["ifrs_statement"],
                "ifrs_section": tr["ifrs_section"],
                "ifrs_line_item": tr["ifrs_line_item"],
                "dot_path": dot_path,
                "match_quality": tr["match_quality"],
            }
        )
    return successes, failures
